### DISCUSSION TITLE
bash script identify: Howto: If value is not set..

### DISCUSSION THREAD

**Discussion by donaufisch66** [ImageMagick/ImageMagick#6273] [Help]
### ImageMagick version

7.0.9.5

### Operating system, version and so on

Linux 5.10.60-qnap x86_64

### Description

Hello,
I write a bash script and use "identify" to get the resolution of an image. A variable should hold the value.
`iRes=$(identify -ping -format '%y' $iFile)`
This works for every file if a resolutin is set. But if the resolution is not set in the picture, I cannot check the variable. I have tried different options.
` if [ -z $iRes ]   if [ $iRes -eq 0 ]  `
All optins to check my variable fail. 
I can do something with all files that have set a resolution . But I need all files that have no resulotion!

What can I do to check for files with no resulotion set? Is it passible to set the value to any string value, to use that as condition for the If ?

my script
`for iFile in $StartSearchDir/*
do 
				iRes=$(identify -ping -format '%y\x' $iFile)
				if	[ -z $iRes ], then 
				echo $iFile
				echo "The variable has not been set"		
			    fi;	
done`

**Comment by fmw42** [comment]
try

```
infile="your_image.suffix"
declare `magick "$infile" -format "xres=%x\n,yres=%y\n" info:`
[ "$xres" = "" ] && xres="xres is not set"
[ "$yres" = "" ] && yres="yres is not set"
echo "xres=$xres"
echo "yres=$yres"

```

**Reply by donaufisch66** [reply]
That doesnt work for me.
I get an error.

Meanwhile I managed is by using `iRes=$(identify -format "%[fx:resolution.x] \n" $1);`. This works but not that fast as `-ping` could do.

**Comment by snibgo** [comment]
> I get an error.

It would help if you said what the error was. Perhaps it was something like:
```
bash: declare: `,yres=118.10999999999999943': not a valid identifier
```
The problem is the comma in the format. Remove it.
```
declare `magick "$infile" -format "xres=%x\nyres=%y\n" info:`
```

**Reply by donaufisch66** [reply]
Hello,
the background of my question is, that I have a big picture backup, where photos are mixed up with received Whatsapp pictures. I want to look for these Whatsapp pics in and delete them.
I have just found out, that Whatsapp pics, habe no xResolution or software set in the exif data.  So I want to look for both and values, and in case they are not set, they have to be Whatsapp pictures.

Related to the first script, the Error was: 
`magick: UnableToOpenBlob '/share/alt/a.JPG': No such file or directory @ error/blob.c/OpenBlob/3496.
BASH=/bin/bash
BASH_ARGC=()
BASH_ARGV=()
BASH_LINENO=([0]="0")
BASH_SOURCE=([0]="t1")`
followed by about50 additional lines with system information.


but this was my mistake, because I defined the picture inluding path within double quotes.
without path an quotes, I also get "not a valid identifier" with the first code. 

Without the comma the script generally works! thank you.

But the problem is, that magick calculates the dpi value. So I get "72" for all files, even if xresolution is not defined in the file.


When I use `identify -ping -format "%[fx:resolution.x] \n" test.jpg ` the  code gives "0" in case the xResolution is not set in a file.

With this line, I can identify pics with no xResolution set. Additional I want to check is there is a Software set in the picture.
If there is no software and no xResolution, I have probably identified a Whatsapp pic in the picture backup.

There are a lot of pics in that backup and my script takes a verry long time to check each file. I would perfer to use `identify -ping` to seach for Whatsapp pics in that backup. "ping" takes just a fraction of that time.
For each result I make a Softlink to that pic in another folder. Then I can check the pics and in case they are all Whatsapp pictures, I will substitute `ln`by `rm` to get rid of that pics.

In my script I need to use identify and exiftool. So I have to check each file twice. I wasnt able to  assign two values to two variables. The problem is, that magick calculates the value or gives an error! But I need any definated state to check the value with an if.

`
#!/bin/bash

searchDir='/share/alt/test'
destDir='/share/alt/dest'

find $searchDir -iname "*.JPG"  -exec bash -c ' 
xRes1=$(identify -ping -format "%[fx:resolution.x] \n" "$1");
if [ "$xRes1" -eq "0" ]; then
	Soft=$(exiftool -s -s -s -f  -"Software" $1);
	$xSoft;
	xString="${Soft:0:7}"; 
	if  [ "$Soft" = "-" ] || [ "$xString" = "Android" ] ; then 
		echo "WhatsApp PIC found";
		ln -s $1 $3;
		fi;
	fi;

' bash {} $searchDir $destDir \;
`

  
My script works but is verry slow. Maybe you have a suggetion, how to speed it up

**Comment by snibgo** [comment]
Why are you using fx? You don't need it. You can use `-format "%[resolution.x] \n"`.

You say your script is very slow. Which part is slow? If the IM command is slower than the exiftool command, then reversing them may improve the speed.

**Comment by fmw42** [comment]
Don't test if resolution is 0.  Test if empty value returned as I showed above. Add -ping to make it faster.  Use Magick identify if you want rather than Magick.

```
infile="your_image.suffix"
declare `magick -ping "$infile" -format "xres=%x\n,yres=%y\n" info:`
[ "$xres" = "" ] && xres="xres is not set"
[ "$yres" = "" ] && yres="yres is not set"
echo "xres=$xres"
echo "yres=$yres"

```

If you do not like declare, then find each value separately

```
infile="your_image.suffix"
xres=`magick -ping "$infile" -format "%x" info:`
yres=`magick -ping "$infile" -format "%y" info:`
[ "$xres" = "" ] && xres="xres is not set"
[ "$yres" = "" ] && yres="yres is not set"
echo "xres=$xres"
echo "yres=$yres"

```

**Reply by donaufisch66** [reply]
I have tested different commands.
```
12 image files read

time identify -ping -format '%x %y \n ' *.*
real    0m0.024s
user    0m0.016s
sys     0m0.008s
time exiftool -s -s -s -f  -"Software" *.*
real    0m0.642s
user    0m0.613s
sys     0m0.029s
time exiftool -s -s -s -f  -"xResolution" *.*
real    0m0.644s
user    0m0.621s
sys     0m0.023s
identify -ping -format "%[EXIF:Software] %[fx:resolution.x] \n" *.*
real    0m1.969s
user    0m1.886s
sys     0m0.071s
for file in /share/alt/a/* ; do identify -ping -format "%[EXIF:Software] %[fx:resolution.x] \n" $file; done
real    0m2.200s
user    0m1.922s
sys     0m0.277s
time identify -ping *.*
real    0m0.025s
user    0m0.020s
sys     0m0.005s

```
Far and away the fastest command is `identify -ping`. But the given  values do not help. 

The script is slow, because it has to check lots of pictures. The backup  is my sons Smartphone backup. He saved all pictures together with all whatsapp pictures and all are renamend. So the only chance to find whatsapp pictures automatically is by searching exif values.
Whatsapp rejects almost all exiff data from pictures.  
Meanwhile I managed to assign two variables in one command. This script works a little bit faster even if the `ping` option is not used or can not be used.. 

```
find $searchDir -iname "*.JPG"  -exec bash -c \' 

read xRes iSoft < <(identify -ping -format "%[resolution.x] %[EXIF:Software] \n" $1 2>NUL );
echo $1;
	if  [ "$xRes" = "0" ] && [ "$iSoft" = "" ] ; then
	echo "WhatsApp PIC found";
	ln -s $1 $3;	
	fi;
\' bash {} $searchDir $destDir \;

```
